### PR TITLE
Removed references to the Number of servers slider.

### PR DESCRIPTION
**Pull request by danielimprobable**
**Contributions**: We are not currently taking public contributions - see our [contributions](CONTRIBUTING.md) policy. However, we are accepting issues and we do want your [feedback](../README.md#give-us-feedback).

-------

#### Description

This PR updated the GDK to no longer reference the removed `Number of servers`. This was removed in the related engine PR: https://github.com/improbableio/UnrealEngine/pull/157.

In addition to this, it also adds the propagation of the GDK settings to the LevelEditorPlaySettings.

#### Release note
Changelog.md has been updated with a new feature entry.

#### Tests
Tested manually. 

#### Documentation

No documentation included with the PR this will come at a later stage.

#### Primary reviewers
@mattyoung-improbable @Vatyx

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDKEditorToolbar/Private/SpatialGDKEditorToolbar.cpp [hunk 1]
```diff
@@ -359,9 +359,20 @@ bool FSpatialGDKEditorToolbarModule::ValidateGeneratedLaunchConfig() const
...
+	bool bLaunchingMultipleServerWorkers = false;
+	if(const int32* ServerCount = LevelEditorPlaySettings->WorkerTypesToLaunch.Find(SpatialConstants::ServerWorkerType))
+	{
+		if (*ServerCount > 1)
+		{
+			bLaunchingMultipleServerWorkers = true;
+		}
+	}
+
+	bLaunchingMultipleServerWorkers = bLaunchingMultipleServerWorkers || LevelEditorPlaySettings->WorkerTypesToLaunch.Num() > 1;
```
Comment (tenevdev) [inline SpatialGDK/Source/SpatialGDKEditorToolbar/Private/SpatialGDKEditorToolbar.cpp]: Entries in `WorkerTypesToLaunch` could have a value of 0 (I think) which would make this flag a false positive. I suggest replacing this and the above conditional with a function that just checks against the sum of all values in `WorkerTypesToLaunch`.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/EditorWorkerController.h [hunk 2]
```diff
@@ -45,20 +45,30 @@ struct EditorWorkerController
 		FSpatialGDKEditorToolbarModule& Toolbar = FModuleManager::GetModuleChecked<FSpatialGDKEditorToolbarModule>("SpatialGDKEditorToolbar");
 		SpatialShutdownHandle = Toolbar.OnSpatialShutdown.AddRaw(this, &EditorWorkerController::OnSpatialShutdown);
 
-		int32 PlayNumberOfServers;
-		GetDefault<ULevelEditorPlaySettings>()->GetPlayNumberOfServers(PlayNumberOfServers);
-
-		WorkerIds.SetNum(PlayNumberOfServers);
-		for (int i = 0; i < PlayNumberOfServers; ++i)
+		const TMap<FString, int32>& WorkerTypesToLaunch = GetDefault<ULevelEditorPlaySettings>()->WorkerTypesToLaunch;
```
Comment (tenevdev) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/EditorWorkerController.h]: Is it possible to get the version of this from launch config that uses `FWorkerTypeLaunchSection` which is a nicer type?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/EditorWorkerController.h [hunk 3]
```diff
@@ -45,20 +45,30 @@ struct EditorWorkerController
...
+		int32 WorkerCount = 0;
+		for (const TPair<FString, int32>& WorkerType : WorkerTypesToLaunch)
 		{
-			FString NewWorkerId = WorkerType + FGuid::NewGuid().ToString();
+			WorkerCount += WorkerType.Value;
+		}
 
-			if (!WorkerIds[i].IsEmpty() && SecondsSinceLastSession < WorkerReplaceThresholdSeconds)
+		WorkerIds.SetNum(WorkerCount);
```
Comment (tenevdev) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/EditorWorkerController.h]: Why do you resize this array and not the `ReplaceProcesses` one?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/EditorWorkerController.h (hunk 2 above)
Comment (danielimprobable) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/EditorWorkerController.h]: Yes I think we will be able to, will update

File: SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/EditorWorkerController.h [hunk 4]
```diff
@@ -45,20 +45,30 @@ struct EditorWorkerController
...
+
+		int32 WorkerIdIndex = 0;
+		for (const TPair<FString, int32>& WorkerType : WorkerTypesToLaunch)
+		{
+			for (int i = 0; i < WorkerType.Value; ++i)
 			{
-				ReplaceProcesses.Add(ReplaceWorker(WorkerIds[i], NewWorkerId));
-			}
+				const FString NewWorkerId = WorkerType.Key + FGuid::NewGuid().ToString();
```
Comment (tenevdev) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/EditorWorkerController.h]: How about changing this to use `i` instead of a guid?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/EditorWorkerController.h [hunk 5]
```diff
@@ -45,20 +45,30 @@ struct EditorWorkerController
...
+			WorkerCount += WorkerType.Value;
+		}
 
-			if (!WorkerIds[i].IsEmpty() && SecondsSinceLastSession < WorkerReplaceThresholdSeconds)
+		WorkerIds.SetNum(WorkerCount);
+
+		int32 WorkerIdIndex = 0;
+		for (const TPair<FString, int32>& WorkerType : WorkerTypesToLaunch)
+		{
+			for (int i = 0; i < WorkerType.Value; ++i)
```
Comment (mattyoung-improbable) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/EditorWorkerController.h]:
```suggestion
			for (int i = 0; i < WorkerType.Value; i++)
```

File: SpatialGDK/Source/SpatialGDK/Private/SpatialGDKSettings.cpp [hunk 6]
```diff
@@ -84,6 +85,11 @@ void USpatialGDKSettings::PostEditChangeProperty(FPropertyChangedEvent& Property
 		ValidateOffloadingSettings();
 	}
 
+	if (PropertyName == GET_MEMBER_NAME_CHECKED(USpatialGDKSettings, bEnableOffloading))
```
Comment (mattyoung-improbable) [inline SpatialGDK/Source/SpatialGDK/Private/SpatialGDKSettings.cpp]: Reminder to remove this if we decide to remove the toggle in ULevelEditorPlaySettings.

File: SpatialGDK/Source/SpatialGDKEditorToolbar/Private/SpatialGDKEditorToolbar.cpp [hunk 7]
```diff
@@ -359,9 +359,20 @@ bool FSpatialGDKEditorToolbarModule::ValidateGeneratedLaunchConfig() const
...
+	
+	if (!SpatialGDKRuntimeSettings->bEnableHandover && bLaunchingMultipleServerWorkers)
```
Comment (mattyoung-improbable) [inline SpatialGDK/Source/SpatialGDKEditorToolbar/Private/SpatialGDKEditorToolbar.cpp]: I'm not sure if this check is valid anymore since pure offloading doesn't need handover. Do you know if handover is required for offloading to worker? If not we could modify this to check if there is >1 of any single worker type.

File: CHANGELOG.md [hunk 8]
```diff
@@ -21,6 +21,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Added runtime option to auto spawn the ASpatialMetricsDisplay for remote debugging of server metrics - bEnableMetricsDisplay.
 - Added runtime option to batch spatial position updates to runtime.
 - Added runtime option to pack unreliable RPCs from the same frame in a single component update to save bandwidth - bPackUnreliableRPCs.
+- Removed the references to the `Number of servers` slider in the Play in editor drop-down menu.
```
Comment (alastairdglennie) [inline CHANGELOG.md]: Can we be slightly more helpful in saying where this information or similar will be exposed instead?

File: SpatialGDK/Source/SpatialGDK/Private/SpatialGDKSettings.cpp [hunk 9]
```diff
@@ -6,8 +6,9 @@
 #include "GameFramework/Actor.h"
 
 #if WITH_EDITOR
-#include "Modules/ModuleManager.h"
 #include "ISettingsModule.h"
+#include "Modules/ModuleManager.h"
+#include "Settings/LevelEditorPlaySettings.h"
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/SpatialGDKSettings.cpp]: was this an IWYU? I can't see any LevelEditor references within this file from before and you've made no other changes?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/EditorWorkerController.h (hunk 5 above)
Comment (alastairdglennie) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/EditorWorkerController.h]: why???

File: SpatialGDK/Source/SpatialGDK/Private/SpatialGDKSettings.cpp (hunk 9 above)
Comment (danielimprobable) [reply SpatialGDK/Source/SpatialGDK/Private/SpatialGDKSettings.cpp]: will revert this change now

File: SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorSettings.cpp [hunk 10]
```diff
@@ -52,6 +52,7 @@ void USpatialGDKEditorSettings::PostInitProperties()
 	PlayInSettings->SaveConfig();
 
 	SetRuntimeWorkerTypes();
+	SetLevelEditorPlaySettingsWorkerTypes();
```
Comment (tenevdev) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorSettings.cpp]: I think this also needs to be called every time there's a change to the LaunchConfigDesc property.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/EditorWorkerController.h [hunk 11]
```diff
@@ -45,20 +46,26 @@ struct EditorWorkerController
 		FSpatialGDKEditorToolbarModule& Toolbar = FModuleManager::GetModuleChecked<FSpatialGDKEditorToolbarModule>("SpatialGDKEditorToolbar");
 		SpatialShutdownHandle = Toolbar.OnSpatialShutdown.AddRaw(this, &EditorWorkerController::OnSpatialShutdown);
 
-		int32 PlayNumberOfServers;
-		GetDefault<ULevelEditorPlaySettings>()->GetPlayNumberOfServers(PlayNumberOfServers);
+		const ULevelEditorPlaySettings* LevelEditorPlaySettings = GetDefault<ULevelEditorPlaySettings>();
+		const int32 WorkerCount = LevelEditorPlaySettings->GetTotalServerWorkerCount();
```
Comment (danielimprobable) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/EditorWorkerController.h]: @tenevdev introduced this function, let me know if it will do what you need for your debugging settings?

Comment (tenevdev) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/EditorWorkerController.h]: Yes, I'll use it but left a comment on the engine PR to clarify that this is not the count of connected servers, just the setting of how many to launch from editor.

File: CHANGELOG.md [hunk 12]
```diff
@@ -103,7 +104,7 @@ In addition to all of the updates from Improbable, this release includes x impro
 
 ## [`0.3.0`](https://github.com/spatialos/UnrealGDK/releases/tag/0.3.0) - 2019-04-04
 
-### New Known Issues: 
+### New Known Issues:
```
Comment (alastairdglennie) [inline CHANGELOG.md]: Bunch of random space deletions in this file?

File: SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorSettings.cpp [hunk 13]
```diff
@@ -116,3 +118,13 @@ void USpatialGDKEditorSettings::SafetyCheckSpatialOSDirectoryPaths()
 		SaveConfig(CPF_Config, *GetDefaultConfigFilename());
 	}
 }
+
+void USpatialGDKEditorSettings::SetLevelEditorPlaySettingsWorkerTypes()
+{
+	ULevelEditorPlaySettings* PlayInSettings = GetMutableDefault<ULevelEditorPlaySettings>();
+
+	for (const FWorkerTypeLaunchSection WorkerLaunch : LaunchConfigDesc.Workers)
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorSettings.cpp]: should this be a &?

File: CHANGELOG.md (hunk 12 above)
Comment (danielimprobable) [reply CHANGELOG.md]: I think it is my editor config that trims the files by default

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 14]
```diff
@@ -171,11 +171,11 @@ void USpatialNetDriver::InitiateConnectionToSpatialOS(const FURL& URL)
 		Connection->LocatorConfig.PlayerIdentityToken = URL.GetOption(TEXT("playeridentity="), TEXT(""));
 		Connection->LocatorConfig.LoginToken = URL.GetOption(TEXT("login="), TEXT(""));
 		Connection->LocatorConfig.UseExternalIp = true;
-		Connection->LocatorConfig.WorkerType = GameInstance->GetSpatialWorkerType();
+		Connection->LocatorConfig.WorkerType = GameInstance->GetSpatialWorkerType().ToString();
 	}
 	else // Using Receptionist
 	{
-		Connection->ReceptionistConfig.WorkerType = GameInstance->GetSpatialWorkerType();
+		Connection->ReceptionistConfig.WorkerType = GameInstance->GetSpatialWorkerType().ToString();
```
Comment (danielimprobable) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: Maybe the `WorkerType` in the Receptionist and locator configs should be FNames as well? what do you guys think cc @mattyoung-improbable @Vatyx @tenevdev @alastairdglennie

Comment (tenevdev) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: I don't see a reason why they should be FName, the one in the game instance is specifically because we expect it to be used in many equality checks.

Comment (alastairdglennie) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: Maybe worth seeing just where `ReceptionistConfig.WorkerType` is used, could be similarly many equality checks

Comment (Vatyx) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: I agree with Tencho that they don't need to be FNames.

File: SpatialGDK/Source/SpatialGDK/Private/SpatialGDKSettings.cpp [hunk 15]
```diff
@@ -3,6 +3,7 @@
 #include "SpatialGDKSettings.h"
 #include "Misc/MessageDialog.h"
 #include "Misc/CommandLine.h"
+#include "GameFramework/Actor.h"
```
Comment (Vatyx) [inline SpatialGDK/Source/SpatialGDK/Private/SpatialGDKSettings.cpp]: Alphabetize please

File: SpatialGDK/Source/SpatialGDK/Private/SpatialGDKSettings.cpp [hunk 16]
```diff
@@ -1,6 +1,7 @@
 // Copyright (c) Improbable Worlds Ltd, All Rights Reserved
 
 #include "SpatialGDKSettings.h"
+#include "GameFramework/Actor.h"
```
Comment (tenevdev) [inline SpatialGDK/Source/SpatialGDK/Private/SpatialGDKSettings.cpp]: Why is this needed?

File: SpatialGDK/Source/SpatialGDK/Private/SpatialGDKSettings.cpp [hunk 17]
```diff
@@ -37,4 +38,27 @@ void USpatialGDKSettings::PostInitProperties()
...
+void USpatialGDKSettings::PostEditChangeProperty(struct FPropertyChangedEvent& PropertyChangedEvent)
+{
+	Super::PostEditChangeProperty(PropertyChangedEvent);
+
+	// Use MemberProperty here so we report the correct member name for nested changes
+	const FName Name = (PropertyChangedEvent.MemberProperty != nullptr) ? PropertyChangedEvent.MemberProperty->GetFName() : NAME_None;
+
+	if (Name == GET_MEMBER_NAME_CHECKED(USpatialGDKSettings, bEnableOffloading))
+	{
+		ULevelEditorPlaySettings* PlayInSettings = GetMutableDefault<ULevelEditorPlaySettings>();
```
Comment (tenevdev) [inline SpatialGDK/Source/SpatialGDK/Private/SpatialGDKSettings.cpp]: I would have written this as a one-liner but this is also fine.

File: SpatialGDK/Source/SpatialGDKServices/Private/Interop/Connection/EditorWorkerController.cpp [hunk 18]
```diff
@@ -37,20 +37,26 @@ struct EditorWorkerController
 
 		PIEEndHandle = FEditorDelegates::PrePIEEnded.AddRaw(this, &EditorWorkerController::OnPrePIEEnded);
 
-		int32 PlayNumberOfServers;
-		GetDefault<ULevelEditorPlaySettings>()->GetPlayNumberOfServers(PlayNumberOfServers);
+		const ULevelEditorPlaySettings* LevelEditorPlaySettings = GetDefault<ULevelEditorPlaySettings>();
+		const int32 WorkerCount = LevelEditorPlaySettings->GetTotalPIEServerWorkerCount();
+		WorkerIds.SetNum(WorkerCount);
+		ReplaceProcesses.SetNum(WorkerCount);
```
[the remark this answers is not in the document]
Comment (alastairdglennie) [reply SpatialGDK/Source/SpatialGDKServices/Private/Interop/Connection/EditorWorkerController.cpp]: True it might not, but I would argue this is the best thing to do since it's the minimum upper bound preventing reallocation. 

Also just found a redundant `ReplaceProcesses.empty()` at the start of the method which I've now deleted.
